// @arbitrum/sdk/dist/lib/abi/factories/IEthBridgeCreator__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IEthBridgeCreator__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "bridgeTemplate",
        outputs: [
            {
                internalType: "contract IBridge",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "adminProxy",
                type: "address",
            },
            {
                internalType: "address",
                name: "rollup",
                type: "address",
            },
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "delayBlocks",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "futureBlocks",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "delaySeconds",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "futureSeconds",
                        type: "uint256",
                    },
                ],
                internalType: "struct ISequencerInbox.MaxTimeVariation",
                name: "maxTimeVariation",
                type: "tuple",
            },
        ],
        name: "createBridge",
        outputs: [
            {
                internalType: "contract IBridge",
                name: "",
                type: "address",
            },
            {
                internalType: "contract SequencerInbox",
                name: "",
                type: "address",
            },
            {
                internalType: "contract IInboxBase",
                name: "",
                type: "address",
            },
            {
                internalType: "contract IRollupEventInbox",
                name: "",
                type: "address",
            },
            {
                internalType: "contract Outbox",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "inboxTemplate",
        outputs: [
            {
                internalType: "contract IInboxBase",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "outboxTemplate",
        outputs: [
            {
                internalType: "contract Outbox",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "rollupEventInboxTemplate",
        outputs: [
            {
                internalType: "contract IRollupEventInbox",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "sequencerInboxTemplate",
        outputs: [
            {
                internalType: "contract SequencerInbox",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_bridgeTemplate",
                type: "address",
            },
            {
                internalType: "address",
                name: "_sequencerInboxTemplate",
                type: "address",
            },
            {
                internalType: "address",
                name: "_inboxTemplate",
                type: "address",
            },
            {
                internalType: "address",
                name: "_rollupEventInboxTemplate",
                type: "address",
            },
            {
                internalType: "address",
                name: "_outboxTemplate",
                type: "address",
            },
        ],
        name: "updateTemplates",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IEthBridgeCreator__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IEthBridgeCreator__factory = IEthBridgeCreator__factory;
IEthBridgeCreator__factory.abi = _abi;
