// @arbitrum/sdk/dist/lib/abi/factories/IRollupCreator__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IRollupCreator__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "contract IBridgeCreator",
                name: "_bridgeCreator",
                type: "address",
            },
            {
                internalType: "contract IOneStepProofEntry",
                name: "_osp",
                type: "address",
            },
            {
                internalType: "contract IChallengeManager",
                name: "_challengeManagerLogic",
                type: "address",
            },
            {
                internalType: "contract IRollupAdmin",
                name: "_rollupAdminLogic",
                type: "address",
            },
            {
                internalType: "contract IRollupUser",
                name: "_rollupUserLogic",
                type: "address",
            },
            {
                internalType: "address",
                name: "_validatorUtils",
                type: "address",
            },
            {
                internalType: "address",
                name: "_validatorWalletCreator",
                type: "address",
            },
        ],
        name: "setTemplates",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IRollupCreator__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IRollupCreator__factory = IRollupCreator__factory;
IRollupCreator__factory.abi = _abi;
